File: test_ws_analysis.py
#!/usr/bin/env python3
"""
WebSocket Quote Analysis Tool
Analyzes ticks from ws://localhost:7999/ws to determine if they are real or simulated
"""

import asyncio
import json
import websockets
import sys
from datetime import datetime
from collections import defaultdict
from statistics import mean, stdev

# Configuration
WS_URL = "ws://localhost:7999/ws"
CAPTURE_COUNT = 20
TIMEOUT = 30

# Data collection
captured_ticks = defaultdict(list)
total_tick_count = 0
start_time = datetime.now()

# Analysis results
analysis = {
    'lp_sources': set(),
    'symbols': set(),
    'evidence': []
}

def analyze_lp(ticks):
    """Analyze LP sources"""
    print('\n📊 LP SOURCE ANALYSIS:')
    print('═' * 60)

    lp_counts = {}
    for tick in ticks:
        lp = tick.get('lp', 'UNKNOWN')
        lp_counts[lp] = lp_counts.get(lp, 0) + 1
        analysis['lp_sources'].add(lp)

    for lp, count in lp_counts.items():
        percentage = (count / len(ticks)) * 100
        print(f"  {lp}: {count} ticks ({percentage:.1f}%)")

        if lp == 'SIMULATED':
            analysis['evidence'].append('❌ LP field shows "SIMULATED" - clear simulation marker')
        elif lp == 'YOFX':
            analysis['evidence'].append('✅ LP field shows "YOFX" - indicates real FIX gateway data')

def analyze_price_patterns(ticks):
    """Analyze price movement patterns"""
    print('\n📈 PRICE PATTERN ANALYSIS:')
    print('═' * 60)

    by_symbol = defaultdict(list)
    for tick in ticks:
        by_symbol[tick['symbol']].append(tick)

    for symbol, symbol_ticks in by_symbol.items():
        if len(symbol_ticks) < 3:
            continue

        # Calculate price changes
        price_changes = []
        for i in range(1, len(symbol_ticks)):
            change = symbol_ticks[i]['bid'] - symbol_ticks[i-1]['bid']
            price_changes.append(change)

        # Check for regular patterns
        unique_changes = len(set(round(c, 6) for c in price_changes))
        regularity_ratio = unique_changes / len(price_changes) if price_changes else 0

        print(f"  {symbol}:")
        print(f"    - Price changes: {len(price_changes)}")
        print(f"    - Unique changes: {unique_changes}")
        print(f"    - Regularity ratio: {regularity_ratio:.3f} (lower = more regular)")

        if regularity_ratio < 0.5:
            analysis['evidence'].append(f"⚠️  {symbol}: Low regularity ratio ({regularity_ratio:.3f}) suggests simulated data")
        else:
            analysis['evidence'].append(f"✅ {symbol}: High price variation ({regularity_ratio:.3f}) suggests real market data")

        # Sample changes
        sample_changes = [f"{c:.6f}" for c in price_changes[:5]]
        print(f"    - Sample changes: [{', '.join(sample_changes)}]")

def analyze_timestamps(ticks):
    """Analyze timestamp patterns"""
    print('\n⏰ TIMESTAMP ANALYSIS:')
    print('═' * 60)

    timestamps = sorted([t['timestamp'] for t in ticks])
    gaps = [timestamps[i] - timestamps[i-1] for i in range(1, len(timestamps))]

    if not gaps:
        print("  Not enough data for timestamp analysis")
        return

    avg_gap = mean(gaps)
    unique_gaps = len(set(gaps))
    gap_regularity = unique_gaps / len(gaps)

    print(f"  Average gap: {avg_gap:.2f}s")
    print(f"  Unique gaps: {unique_gaps} / {len(gaps)}")
    print(f"  Gap regularity: {gap_regularity:.3f}")

    # Check for regular intervals
    if gap_regularity < 0.3 and avg_gap < 1:
        analysis['evidence'].append(f"⚠️  Very regular timestamp intervals ({avg_gap:.2f}s) suggests simulation")
    else:
        analysis['evidence'].append("✅ Irregular timestamp intervals suggest real market data")

    # Sample gaps
    sample_gaps = [f"{g:.2f}" for g in gaps[:10]]
    print(f"  Sample gaps (s): [{', '.join(sample_gaps)}]")

def analyze_spread(ticks):
    """Analyze spread consistency"""
    print('\n💰 SPREAD ANALYSIS:')
    print('═' * 60)

    by_symbol = defaultdict(list)
    for tick in ticks:
        by_symbol[tick['symbol']].append(tick)

    for symbol, symbol_ticks in by_symbol.items():
        spreads = [round(t['ask'] - t['bid'], 6) for t in symbol_ticks]
        unique_spreads = len(set(spreads))
        spread_consistency = 1 - (unique_spreads / len(spreads))

        print(f"  {symbol}:")
        print(f"    - Total ticks: {len(spreads)}")
        print(f"    - Unique spreads: {unique_spreads}")
        print(f"    - Consistency: {spread_consistency * 100:.1f}%")

        if spread_consistency > 0.95:
            analysis['evidence'].append(f"⚠️  {symbol}: Perfectly consistent spread ({spread_consistency * 100:.1f}%) suggests simulation")
        else:
            analysis['evidence'].append(f"✅ {symbol}: Variable spread suggests real market conditions")

        # Sample spreads
        sample_spreads = [f"{s:.6f}" for s in list(set(spreads))[:5]]
        print(f"    - Sample spreads: [{', '.join(sample_spreads)}]")

def generate_verdict():
    """Generate final verdict"""
    print('\n\n' + '═' * 60)
    print('🔍 FINAL VERDICT')
    print('═' * 60)

    # Count indicators
    simulated_indicators = sum(1 for e in analysis['evidence'] if '⚠️' in e or '❌' in e)
    real_indicators = sum(1 for e in analysis['evidence'] if '✅' in e)

    print('\nEVIDENCE SUMMARY:')
    for evidence in analysis['evidence']:
        print(f"  {evidence}")

    print('\n' + '─' * 60)
    print(f"Simulated indicators: {simulated_indicators}")
    print(f"Real data indicators: {real_indicators}")
    print('─' * 60)

    # Determine verdict
    if 'SIMULATED' in analysis['lp_sources']:
        verdict = 'SIMULATED'
        print('\n🎭 VERDICT: SIMULATED DATA')
        print('   The LP field explicitly shows "SIMULATED"')
    elif 'YOFX' in analysis['lp_sources'] and real_indicators > simulated_indicators:
        verdict = 'REAL'
        print('\n✅ VERDICT: REAL MARKET DATA')
        print('   LP field shows "YOFX" (FIX gateway)')
        print('   Price patterns and timestamps show real market characteristics')
    elif simulated_indicators > real_indicators:
        verdict = 'LIKELY_SIMULATED'
        print('\n⚠️  VERDICT: LIKELY SIMULATED')
        print('   Patterns suggest simulation despite LP field')
    else:
        verdict = 'INCONCLUSIVE'
        print('\n❓ VERDICT: INCONCLUSIVE')
        print('   Mixed indicators - need more data')

    duration = (datetime.now() - start_time).total_seconds()
    print('\n' + '═' * 60)
    print(f"Total ticks analyzed: {total_tick_count}")
    print(f"Symbols detected: {', '.join(sorted(analysis['symbols']))}")
    print(f"Duration: {duration:.1f}s")
    print('═' * 60)

async def capture_and_analyze():
    """Main capture and analysis function"""
    global total_tick_count

    print('═' * 60)
    print('🔌 WebSocket Quote Analysis Tool')
    print('═' * 60)
    print(f"Connecting to: {WS_URL}")
    print(f"Target: Capture {CAPTURE_COUNT} ticks per symbol")
    print(f"Timeout: {TIMEOUT}s")
    print('═' * 60)

    try:
        async with websockets.connect(WS_URL, ping_interval=None) as websocket:
            print('✅ Connected to WebSocket')
            print('📊 Capturing ticks...\n')

            start = datetime.now()

            while True:
                # Check timeout
                if (datetime.now() - start).total_seconds() > TIMEOUT:
                    print('\n⏰ Analysis timeout reached')
                    break

                try:
                    # Receive message with timeout
                    message = await asyncio.wait_for(websocket.recv(), timeout=5.0)
                    tick = json.loads(message)

                    if tick.get('type') == 'tick':
                        total_tick_count += 1
                        symbol = tick['symbol']
                        analysis['symbols'].add(symbol)

                        # Store tick
                        if len(captured_ticks[symbol]) < CAPTURE_COUNT:
                            captured_ticks[symbol].append(tick)

                        # Log first few ticks
                        if total_tick_count <= 5:
                            timestamp = datetime.fromtimestamp(tick['timestamp']).isoformat()
                            print(f"[{total_tick_count}] {symbol} | Bid: {tick['bid']:.5f} | Ask: {tick['ask']:.5f} | LP: {tick.get('lp', 'N/A')} | Time: {timestamp}")

                        # Check if we have enough data
                        symbols_with_enough = sum(1 for ticks in captured_ticks.values() if len(ticks) >= CAPTURE_COUNT)
                        if symbols_with_enough >= min(3, len(analysis['symbols'])) and total_tick_count >= CAPTURE_COUNT:
                            break

                except asyncio.TimeoutError:
                    continue
                except json.JSONDecodeError as e:
                    print(f"Error parsing tick: {e}")
                    continue

    except Exception as e:
        print(f"❌ Connection error: {e}")
        sys.exit(1)

    print('\n🔌 WebSocket disconnected')

    # Perform analysis
    all_ticks = [tick for ticks in captured_ticks.values() for tick in ticks]

    if not all_ticks:
        print('\n❌ No ticks captured. Server may not be sending data.')
        sys.exit(1)

    analyze_lp(all_ticks)
    analyze_price_patterns(all_ticks)
    analyze_timestamps(all_ticks)
    analyze_spread(all_ticks)
    generate_verdict()

if __name__ == '__main__':
    try:
        asyncio.run(capture_and_analyze())
    except KeyboardInterrupt:
        print('\n\n⚠️  Analysis interrupted by user')
        sys.exit(0)

File: test_ws_analysis.js
#!/usr/bin/env node
/**
 * WebSocket Quote Analysis Tool
 * Connects to ws://localhost:7999/ws and analyzes ticks to determine if they are real or simulated
 */

const WebSocket = require('ws');

// Configuration
const WS_URL = 'ws://localhost:7999/ws';
const CAPTURE_COUNT = 20; // Number of ticks to capture per symbol
const ANALYSIS_TIMEOUT = 30000; // 30 seconds timeout

// Data collection
const capturedTicks = {};
let totalTickCount = 0;
let startTime = Date.now();

// Analysis results
const analysis = {
    lpSources: new Set(),
    pricePatterns: [],
    timestampGaps: [],
    spreadVariations: [],
    symbols: new Set(),
    verdict: null,
    evidence: []
};

function analyzeLP(ticks) {
    const lpCounts = {};
    ticks.forEach(tick => {
        const lp = tick.lp || 'UNKNOWN';
        lpCounts[lp] = (lpCounts[lp] || 0) + 1;
        analysis.lpSources.add(lp);
    });

    console.log('\n📊 LP SOURCE ANALYSIS:');
    console.log('═'.repeat(60));
    Object.entries(lpCounts).forEach(([lp, count]) => {
        const percentage = ((count / ticks.length) * 100).toFixed(1);
        console.log(`  ${lp}: ${count} ticks (${percentage}%)`);

        if (lp === 'SIMULATED') {
            analysis.evidence.push(`❌ LP field shows "SIMULATED" - clear simulation marker`);
        } else if (lp === 'YOFX') {
            analysis.evidence.push(`✅ LP field shows "YOFX" - indicates real FIX gateway data`);
        }
    });
}

function analyzePricePatterns(ticks) {
    console.log('\n📈 PRICE PATTERN ANALYSIS:');
    console.log('═'.repeat(60));

    // Group by symbol
    const bySymbol = {};
    ticks.forEach(tick => {
        if (!bySymbol[tick.symbol]) bySymbol[tick.symbol] = [];
        bySymbol[tick.symbol].push(tick);
    });

    Object.entries(bySymbol).forEach(([symbol, symbolTicks]) => {
        if (symbolTicks.length < 3) return;

        // Calculate price changes
        const priceChanges = [];
        for (let i = 1; i < symbolTicks.length; i++) {
            const change = symbolTicks[i].bid - symbolTicks[i-1].bid;
            priceChanges.push(change);
        }

        // Check for regular patterns (simulated data often has regular patterns)
        const uniqueChanges = new Set(priceChanges.map(c => c.toFixed(6)));
        const regularityRatio = uniqueChanges.size / priceChanges.length;

        console.log(`  ${symbol}:`);
        console.log(`    - Price changes: ${priceChanges.length}`);
        console.log(`    - Unique changes: ${uniqueChanges.size}`);
        console.log(`    - Regularity ratio: ${regularityRatio.toFixed(3)} (lower = more regular)`);

        if (regularityRatio < 0.5) {
            analysis.evidence.push(`⚠️  ${symbol}: Low regularity ratio (${regularityRatio.toFixed(3)}) suggests simulated data`);
        } else {
            analysis.evidence.push(`✅ ${symbol}: High price variation (${regularityRatio.toFixed(3)}) suggests real market data`);
        }

        // Sample of actual changes
        const sampleChanges = priceChanges.slice(0, 5).map(c => c.toFixed(6));
        console.log(`    - Sample changes: [${sampleChanges.join(', ')}]`);
    });
}

function analyzeTimestamps(ticks) {
    console.log('\n⏰ TIMESTAMP ANALYSIS:');
    console.log('═'.repeat(60));

    const timestamps = ticks.map(t => t.timestamp).sort((a, b) => a - b);
    const gaps = [];

    for (let i = 1; i < timestamps.length; i++) {
        const gap = timestamps[i] - timestamps[i-1];
        gaps.push(gap);
    }

    // Statistical analysis
    const avgGap = gaps.reduce((a, b) => a + b, 0) / gaps.length;
    const uniqueGaps = new Set(gaps);
    const gapRegularity = uniqueGaps.size / gaps.length;

    console.log(`  Average gap: ${avgGap.toFixed(2)}s`);
    console.log(`  Unique gaps: ${uniqueGaps.size} / ${gaps.length}`);
    console.log(`  Gap regularity: ${gapRegularity.toFixed(3)}`);

    // Check for perfectly regular intervals (sign of simulation)
    if (gapRegularity < 0.3 && avgGap < 1) {
        analysis.evidence.push(`⚠️  Very regular timestamp intervals (${avgGap.toFixed(2)}s) suggests simulation`);
    } else {
        analysis.evidence.push(`✅ Irregular timestamp intervals suggest real market data`);
    }

    // Sample gaps
    const sampleGaps = gaps.slice(0, 10).map(g => g.toFixed(2));
    console.log(`  Sample gaps (s): [${sampleGaps.join(', ')}]`);
}

function analyzeSpread(ticks) {
    console.log('\n💰 SPREAD ANALYSIS:');
    console.log('═'.repeat(60));

    const bySymbol = {};
    ticks.forEach(tick => {
        if (!bySymbol[tick.symbol]) bySymbol[tick.symbol] = [];
        bySymbol[tick.symbol].push(tick);
    });

    Object.entries(bySymbol).forEach(([symbol, symbolTicks]) => {
        const spreads = symbolTicks.map(t => (t.ask - t.bid).toFixed(6));
        const uniqueSpreads = new Set(spreads);
        const spreadConsistency = 1 - (uniqueSpreads.size / spreads.length);

        console.log(`  ${symbol}:`);
        console.log(`    - Total ticks: ${spreads.length}`);
        console.log(`    - Unique spreads: ${uniqueSpreads.size}`);
        console.log(`    - Consistency: ${(spreadConsistency * 100).toFixed(1)}%`);

        if (spreadConsistency > 0.95) {
            analysis.evidence.push(`⚠️  ${symbol}: Perfectly consistent spread (${(spreadConsistency * 100).toFixed(1)}%) suggests simulation`);
        } else {
            analysis.evidence.push(`✅ ${symbol}: Variable spread suggests real market conditions`);
        }

        // Sample spreads
        const sampleSpreads = [...uniqueSpreads].slice(0, 5);
        console.log(`    - Sample spreads: [${sampleSpreads.join(', ')}]`);
    });
}

function generateVerdict() {
    console.log('\n\n' + '═'.repeat(60));
    console.log('🔍 FINAL VERDICT');
    console.log('═'.repeat(60));

    // Count positive and negative indicators
    const simulatedIndicators = analysis.evidence.filter(e => e.includes('⚠️') || e.includes('❌')).length;
    const realIndicators = analysis.evidence.filter(e => e.includes('✅')).length;

    console.log('\nEVIDENCE SUMMARY:');
    analysis.evidence.forEach(evidence => console.log(`  ${evidence}`));

    console.log('\n' + '─'.repeat(60));
    console.log(`Simulated indicators: ${simulatedIndicators}`);
    console.log(`Real data indicators: ${realIndicators}`);
    console.log('─'.repeat(60));

    // Determine verdict
    if (analysis.lpSources.has('SIMULATED')) {
        analysis.verdict = 'SIMULATED';
        console.log('\n🎭 VERDICT: SIMULATED DATA');
        console.log('   The LP field explicitly shows "SIMULATED"');
    } else if (analysis.lpSources.has('YOFX') && realIndicators > simulatedIndicators) {
        analysis.verdict = 'REAL';
        console.log('\n✅ VERDICT: REAL MARKET DATA');
        console.log('   LP field shows "YOFX" (FIX gateway)');
        console.log('   Price patterns and timestamps show real market characteristics');
    } else if (simulatedIndicators > realIndicators) {
        analysis.verdict = 'LIKELY_SIMULATED';
        console.log('\n⚠️  VERDICT: LIKELY SIMULATED');
        console.log('   Patterns suggest simulation despite LP field');
    } else {
        analysis.verdict = 'INCONCLUSIVE';
        console.log('\n❓ VERDICT: INCONCLUSIVE');
        console.log('   Mixed indicators - need more data');
    }

    console.log('\n' + '═'.repeat(60));
    console.log(`Total ticks analyzed: ${totalTickCount}`);
    console.log(`Symbols detected: ${[...analysis.symbols].join(', ')}`);
    console.log(`Duration: ${((Date.now() - startTime) / 1000).toFixed(1)}s`);
    console.log('═'.repeat(60));
}

function main() {
    console.log('═'.repeat(60));
    console.log('🔌 WebSocket Quote Analysis Tool');
    console.log('═'.repeat(60));
    console.log(`Connecting to: ${WS_URL}`);
    console.log(`Target: Capture ${CAPTURE_COUNT} ticks per symbol`);
    console.log(`Timeout: ${ANALYSIS_TIMEOUT/1000}s`);
    console.log('═'.repeat(60));

    const ws = new WebSocket(WS_URL);
    let analysisTimeout;

    ws.on('open', () => {
        console.log('✅ Connected to WebSocket');
        console.log('📊 Capturing ticks...\n');

        // Set timeout for analysis
        analysisTimeout = setTimeout(() => {
            console.log('\n⏰ Analysis timeout reached');
            ws.close();
        }, ANALYSIS_TIMEOUT);
    });

    ws.on('message', (data) => {
        try {
            const tick = JSON.parse(data.toString());

            if (tick.type === 'tick') {
                totalTickCount++;
                analysis.symbols.add(tick.symbol);

                // Store tick
                if (!capturedTicks[tick.symbol]) {
                    capturedTicks[tick.symbol] = [];
                }

                if (capturedTicks[tick.symbol].length < CAPTURE_COUNT) {
                    capturedTicks[tick.symbol].push(tick);

                    // Log first few ticks
                    if (totalTickCount <= 5) {
                        console.log(`[${totalTickCount}] ${tick.symbol} | Bid: ${tick.bid.toFixed(5)} | Ask: ${tick.ask.toFixed(5)} | LP: ${tick.lp || 'N/A'} | Time: ${new Date(tick.timestamp * 1000).toISOString()}`);
                    }
                }

                // Check if we have enough data
                const symbolsWithEnoughData = Object.values(capturedTicks).filter(arr => arr.length >= CAPTURE_COUNT).length;
                if (symbolsWithEnoughData >= Math.min(3, analysis.symbols.size) && totalTickCount >= CAPTURE_COUNT) {
                    clearTimeout(analysisTimeout);
                    ws.close();
                }
            }
        } catch (error) {
            console.error('Error parsing tick:', error.message);
        }
    });

    ws.on('error', (error) => {
        console.error('❌ WebSocket error:', error.message);
        process.exit(1);
    });

    ws.on('close', () => {
        console.log('\n🔌 WebSocket disconnected');

        // Perform analysis
        const allTicks = Object.values(capturedTicks).flat();

        if (allTicks.length === 0) {
            console.log('\n❌ No ticks captured. Server may not be running or not sending data.');
            process.exit(1);
        }

        analyzeLP(allTicks);
        analyzePricePatterns(allTicks);
        analyzeTimestamps(allTicks);
        analyzeSpread(allTicks);
        generateVerdict();

        process.exit(0);
    });
}

// Run the analysis
main();
